File: visualization_py/helpers/signal_utils.py
import numpy as np
import math

from helpers.data_loading import Section


def generate_pulse(bpm, window_length, feature_rate):
    freq = bpm / 60
    T = np.arange(0, window_length, 1) / feature_rate
    twoPiT = 2 * math.pi * T
    twoPiFt = freq * twoPiT
    cosine = np.cos(twoPiFt)
    sine = np.sin(twoPiFt)
    return cosine, sine


def pulse_for_section(section: Section, feature_rate, multiple=1):
    offset_rel = section.offset - section.start
    cosine, sine = generate_pulse(section.bpm * multiple,
                                  int((section.end - section.start + offset_rel) * feature_rate), feature_rate)
    offset_samples = int(offset_rel * feature_rate)
    return cosine[offset_samples:], sine[offset_samples:]


def times_from_feature_rate(feature_rate, start, length):
    return [i / feature_rate for i in range(start, start + length)]

File: libtempo/signal_utils.cpp
//
// Created by egordm on 30-8-2018.
//

#include "signal_utils.h"

std::tuple<vec, vec>
libtempo::signal_utils::generate_pulse(double bpm, int window_length, int feature_rate, int shift_by) {
    double freq = bpm / 60.;
    vec T = regspace<vec>(-shift_by, window_length - 1 - shift_by) / feature_rate;
    vec twoPiFt = freq * 2 * M_PI * T;

    return std::make_tuple(cos(twoPiFt), sin(twoPiFt));
}